File: kinv/core/inventory/adopt.py
"""Reading back what the schematic already says about a part.

The other half of writing ``MPN``, ``Manufacturer``, ``Supplier`` and
``Supplier#`` onto the symbols. Nothing here is automatic: reading a file is
not permission to write a catalog that outlives the board, so this produces a
plan, and applying it is a separate, explicit act.
"""

from __future__ import annotations

from typing import Callable, Iterable

from kinv.core.consolidate.findings import FIELD_JOIN
from kinv.core.inventory.resolve import now_iso, part_id
from kinv.core.parse.spec import PART_NUMBER_FIELDS

# What each fact can be called on a symbol. The first name is the one we write.
_MANUFACTURER_FIELDS = ["Manufacturer", "Mfr", "Mfg", "MFR"]
_SUPPLIER_FIELDS = ["Supplier", "Vendor", "Distributor"]
_ORDER_NUMBER_FIELDS = ["Supplier#", "SupplierPN", "Supplier Part Number", "Vendor#"]


def _field_value(fields: dict[str, str], names: list[str]) -> str:
    """One field's value by any of its names — but a merged disagreement is a question, not a value."""
    lookup = {k.lower(): v for k, v in fields.items()}
    for name in names:
        raw = lookup.get(name.lower())
        if raw is None:
            continue
        value = raw.strip()
        if value == "" or FIELD_JOIN.strip() in value:
            continue
        return value
    return ""


def schematic_part(part: dict) -> dict:
    return {
        "mpn": _field_value(part["fields"], PART_NUMBER_FIELDS),
        "manufacturer": _field_value(part["fields"], _MANUFACTURER_FIELDS),
        "supplier": _field_value(part["fields"], _SUPPLIER_FIELDS),
        "orderNumber": _field_value(part["fields"], _ORDER_NUMBER_FIELDS),
    }


def plan_adopt(parts: Iterable[dict], catalog: Iterable[dict], assignments: Iterable[dict]) -> dict:
    """What the board's own fields would add to the catalog. A read, and only a read."""
    by_key = {a["key"]: a for a in assignments}
    by_id = {p["id"]: p for p in catalog}
    silent = 0
    entries: list[dict] = []

    for part in parts:
        said = schematic_part(part)
        # No part number means nothing to hang the rest on.
        if said["mpn"] == "":
            silent += 1
            continue

        entry = {"key": part["key"], "refs": part["refs"], "mpn": said["mpn"], "status": "new", "fills": {}, "disagrees": []}
        assignment = by_key.get(part["key"])
        assigned = None if assignment is None else by_id.get(assignment["partId"])

        if assigned is not None and assigned["id"] != part_id(said["mpn"]):
            entry["status"] = "conflict"
            entry["disagrees"].append({"field": "MPN", "schematic": said["mpn"], "catalog": assigned["mpn"]})
            entries.append(entry)
            continue

        # The catalog row this MPN already has, whether or not this key points at it.
        existing = assigned if assigned is not None else by_id.get(part_id(said["mpn"]))

        def compare(field: str, source: str, had: str | None) -> None:
            if source == "":
                return
            current = (had or "").strip()
            if current == "":
                entry["fills"][field] = source
            elif current != source:
                entry["disagrees"].append({"field": field, "schematic": source, "catalog": current})

        compare("manufacturer", said["manufacturer"], existing.get("manufacturer") if existing else None)
        compare("supplier", said["supplier"], existing.get("supplier") if existing else None)
        compare("orderNumber", said["orderNumber"], existing.get("orderNumber") if existing else None)

        entry["status"] = "new" if assignment is None else ("fills" if entry["fills"] else "agrees")
        entries.append(entry)

    return {"entries": entries, "silent": silent}


def apply_adopt(plan: dict, catalog: list[dict], assignments: list[dict], now: Callable[[], str] = now_iso) -> dict:
    """Applies a plan: detail fills a gap and never overwrites; ``by: "rule"``, because a file said it."""
    next_catalog = list(catalog)
    next_assignments = list(assignments)
    adopted: list[str] = []
    filled: list[str] = []

    for entry in plan["entries"]:
        if entry["status"] in ("conflict", "agrees"):
            continue
        ident = part_id(entry["mpn"])
        if ident == "":
            continue

        existing = next((p for p in next_catalog if p["id"] == ident), None)
        part = dict(existing) if existing is not None else {
            "id": ident,
            "mpn": entry["mpn"],
            "manufacturer": "",
            "firstKey": entry["key"],
            "addedAt": now(),
        }
        for field, value in entry["fills"].items():
            if field in ("manufacturer", "supplier", "orderNumber"):
                part[field] = value

        next_catalog = [*(p for p in next_catalog if p["id"] != ident), part]
        if entry["fills"]:
            filled.append(entry["key"])

        if entry["status"] == "new":
            next_assignments = [
                *(a for a in next_assignments if a["key"] != entry["key"]),
                {"key": entry["key"], "partId": ident, "decidedAt": now(), "by": "rule"},
            ]
            adopted.append(entry["key"])

    return {"catalog": next_catalog, "assignments": next_assignments, "adopted": adopted, "filled": filled}

File: src/core/inventory/adopt.ts
import { fieldJoin, type PartRollup } from "../consolidate/findings.js";
import { partNumberFields } from "../parse/spec.js";
import { partId } from "./resolve.js";
import type { Assignment, CatalogPart } from "./types.js";

/**
 * Reading back what the schematic already says about a part.
 *
 * `kinv fields write` puts `MPN`, `Manufacturer`, `Supplier` and `Supplier#`
 * onto the symbols; this is the other half of that round trip. Without it the
 * catalog is the only place the work exists, and a board handed to someone else
 * — or opened after `~/.kinv` is gone — is back to knowing nothing.
 *
 * It is also what makes copying a symbol in KiCad worth something: the copy
 * carries the fields, so the part it becomes arrives already answered.
 *
 * Nothing here is automatic. Reading a file is not permission to write a
 * catalog that outlives the board, so this produces a plan and the decision to
 * apply it is a separate, explicit act.
 */

/** What each fact can be called on a symbol. The first name is the one we write. */
const manufacturerFields = ["Manufacturer", "Mfr", "Mfg", "MFR"];
const supplierFields = ["Supplier", "Vendor", "Distributor"];
const orderNumberFields = ["Supplier#", "SupplierPN", "Supplier Part Number", "Vendor#"];

/**
 * One field's value, by any of its names, or "" if the schematic does not say.
 *
 * A merged disagreement is not a value: when `100n` and `100nF` roll up into
 * one part and their MPN fields differ, `bomFields` keeps both joined, and
 * `RC0402FR-0710KL · RC0603FR-0710KL` is a question rather than an answer.
 * Importing either spelling would be the tool picking, which is the one thing
 * step 2 exists to stop it doing.
 */
function fieldValue(fields: Record<string, string>, names: readonly string[]): string {
  const lookup = new Map(Object.entries(fields).map(([k, v]) => [k.toLowerCase(), v]));
  for (const name of names) {
    const raw = lookup.get(name.toLowerCase());
    if (raw === undefined) continue;
    const value = raw.trim();
    if (value === "" || value.includes(fieldJoin.trim())) continue;
    return value;
  }
  return "";
}

/** What one symbol's fields claim about the part it is. */
export type SchematicPart = {
  mpn: string;
  manufacturer: string;
  supplier: string;
  orderNumber: string;
};

export function schematicPart(part: PartRollup): SchematicPart {
  return {
    mpn: fieldValue(part.fields, partNumberFields),
    manufacturer: fieldValue(part.fields, manufacturerFields),
    supplier: fieldValue(part.fields, supplierFields),
    orderNumber: fieldValue(part.fields, orderNumberFields),
  };
}

/**
 * What importing one part would do.
 *
 * - `new` — the catalog has no answer for this key, and the schematic does.
 * - `fills` — the same part, with detail the catalog is missing.
 * - `agrees` — nothing to do, and worth saying so.
 * - `conflict` — the schematic names a different part than the catalog. Never
 *   resolved automatically: one of the two is out of date and only you know
 *   which.
 */
export type AdoptStatus = "new" | "fills" | "agrees" | "conflict";

export type AdoptEntry = {
  key: string;
  refs: string[];
  mpn: string;
  status: AdoptStatus;
  /** Catalog fields this would fill in, by name. */
  fills: Record<string, string>;
  /**
   * Where the schematic and the catalog say different things. The catalog
   * keeps its value in every case — a decision you made is not overwritten by
   * a file — and the disagreement is reported rather than swallowed.
   */
  disagrees: { field: string; schematic: string; catalog: string }[];
};

export type AdoptPlan = {
  entries: AdoptEntry[];
  /** Parts whose symbols carry nothing to import. */
  silent: number;
};

/**
 * What the board's own fields would add to the catalog.
 *
 * A read, and only a read: it touches nothing.
 */
export function planAdopt(
  parts: readonly PartRollup[],
  catalog: readonly CatalogPart[],
  assignments: readonly Assignment[],
): AdoptPlan {
  const byKey = new Map(assignments.map((a) => [a.key, a]));
  const byId = new Map(catalog.map((p) => [p.id, p]));

  let silent = 0;
  const entries: AdoptEntry[] = [];

  for (const part of parts) {
    const said = schematicPart(part);
    // No part number means nothing to hang the rest on, the same rule the
    // vendor columns follow: a manufacturer with no MPN is not a part.
    if (said.mpn === "") {
      silent++;
      continue;
    }

    const entry: AdoptEntry = {
      key: part.key,
      refs: part.refs,
      mpn: said.mpn,
      status: "new",
      fills: {},
      disagrees: [],
    };

    const assignment = byKey.get(part.key);
    const assigned = assignment === undefined ? undefined : byId.get(assignment.partId);

    if (assigned !== undefined && assigned.id !== partId(said.mpn)) {
      entry.status = "conflict";
      entry.disagrees.push({ field: "MPN", schematic: said.mpn, catalog: assigned.mpn });
      entries.push(entry);
      continue;
    }

    // The catalog row this MPN already has, whether or not this key points at
    // it: an MPN entered on another board is the same part.
    const existing = assigned ?? byId.get(partId(said.mpn));
    const compare = (field: string, from: string, had: string | undefined): void => {
      if (from === "") return;
      const current = (had ?? "").trim();
      if (current === "") entry.fills[field] = from;
      else if (current !== from) entry.disagrees.push({ field, schematic: from, catalog: current });
    };
    compare("manufacturer", said.manufacturer, existing?.manufacturer);
    compare("supplier", said.supplier, existing?.supplier);
    compare("orderNumber", said.orderNumber, existing?.orderNumber);

    entry.status =
      assignment === undefined ? "new" : Object.keys(entry.fills).length > 0 ? "fills" : "agrees";
    entries.push(entry);
  }

  return { entries, silent };
}

export type AdoptResult = {
  catalog: CatalogPart[];
  assignments: Assignment[];
  /** Keys that gained an assignment they did not have. */
  adopted: string[];
  /** Keys whose catalog part gained a manufacturer, supplier or order number. */
  filled: string[];
};

/**
 * Applies a plan to the catalog and the assignments.
 *
 * The merge is the one `assignPart` makes: typed detail fills a gap and never
 * overwrites. `by: "rule"`, because the tool derived this from a file — you
 * accepted it, but you did not type it, and the audit log should not claim
 * otherwise.
 */
export function applyAdopt(
  plan: AdoptPlan,
  catalog: readonly CatalogPart[],
  assignments: readonly Assignment[],
  now: () => string = () => new Date().toISOString(),
): AdoptResult {
  let nextCatalog = [...catalog];
  let nextAssignments = [...assignments];
  const adopted: string[] = [];
  const filled: string[] = [];

  for (const entry of plan.entries) {
    if (entry.status === "conflict" || entry.status === "agrees") continue;

    const id = partId(entry.mpn);
    if (id === "") continue;

    const existing = nextCatalog.find((p) => p.id === id);
    const part: CatalogPart = {
      ...(existing ?? {
        id,
        mpn: entry.mpn,
        manufacturer: "",
        firstKey: entry.key,
        addedAt: now(),
      }),
    };
    for (const [field, value] of Object.entries(entry.fills)) {
      if (field === "manufacturer") part.manufacturer = value;
      if (field === "supplier") part.supplier = value;
      if (field === "orderNumber") part.orderNumber = value;
    }

    nextCatalog = [...nextCatalog.filter((p) => p.id !== id), part];
    if (Object.keys(entry.fills).length > 0) filled.push(entry.key);

    if (entry.status === "new") {
      nextAssignments = [
        ...nextAssignments.filter((a) => a.key !== entry.key),
        { key: entry.key, partId: id, decidedAt: now(), by: "rule" },
      ];
      adopted.push(entry.key);
    }
  }

  return { catalog: nextCatalog, assignments: nextAssignments, adopted, filled };
}
